mono/metadata/threads.c

```diff
@@ -2012,11 +2015,13 @@ gint32 ves_icall_System_Threading_WaitHandle_WaitOne_internal(HANDLE handle, gin
 	if (mono_thread_current_check_pending_interrupt ())
 		return map_native_wait_result_to_managed (MONO_W32HANDLE_WAIT_RET_FAILED, 0);
 
+	mono_w32handle_ref(handle);
```

Comment:
inc/dec being atomic doesn't protect against use-after-free without locking when reading.

Incrementing the refcount of a handle is only valid in 2 scenarios:
1)you hold the handles lock - or whatever lock protect the table
2)the handle was already inc'd in code leading to the current one.